File: backend/src/main/java/com/mycompany/online_shop_backend/controller/BookController.java
package com.mycompany.online_shop_backend.controller;

import com.mycompany.online_shop_backend.controller.model.response.GetBookByIdResponseDto;
import com.mycompany.online_shop_backend.controller.model.response.GetBooksResponseDto;
import com.mycompany.online_shop_backend.service.BookService;
import com.mycompany.online_shop_backend.service.model.BookDto;
import lombok.RequiredArgsConstructor;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@RestController
@RequiredArgsConstructor
@RequestMapping(produces = MediaType.APPLICATION_JSON_VALUE)
public class BookController {

    private final BookService bookService;

    @GetMapping(path = "/api/v1/books")
    public List<GetBooksResponseDto> getBooks() {
        return bookService.getAllBooks()
                .stream()
                .map(GetBooksResponseDto::toDto)
                .toList();
    }

    @GetMapping(path = "/api/v1/books/{id}")
    public GetBookByIdResponseDto getBookById(@PathVariable("id") long id) {
        BookDto bookDto = bookService.getById(id);
        return GetBookByIdResponseDto.toDto(bookDto);
    }
}

File: book-service/src/main/kotlin/com/example/bookservice/api/rest/BookController.kt
package com.example.bookservice.api.rest

import com.example.bookservice.api.rest.model.Book
import com.example.bookservice.api.rest.model.BookDescription
import com.example.bookservice.api.rest.model.BookSearchRequest
import com.example.bookservice.api.rest.model.CreateBookRequest
import com.example.bookservice.service.BookService
import org.springframework.data.web.PagedModel
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.web.bind.annotation.*
import java.util.*

@RestController
@RequestMapping("/api/v1", produces = [MediaType.APPLICATION_JSON_VALUE])
class BookController(private val bookService: BookService) {

    @PostMapping("/books/search")
    fun books(
        bookRequestParams: BookRequestParams,
        @RequestBody request: BookSearchRequest?
    ): PagedModel<Book> = bookService.getBooks(bookRequestParams, request)

    @GetMapping("/books/{bookId}")
    fun bookById(@PathVariable("bookId") bookId: UUID): Book = bookService.getBookById(bookId)

    @GetMapping("/books/{bookId}/description")
    fun bookDescription(@PathVariable("bookId") bookId: UUID): BookDescription =
        bookService.getBookDescription(bookId)

    @PostMapping("/books", consumes = [MediaType.APPLICATION_JSON_VALUE])
    @ResponseStatus(HttpStatus.CREATED)
    fun createBook(
        @RequestHeader(IDEMPOTENCY_KEY) idempotencyKey: UUID,
        @RequestBody request: CreateBookRequest
    ): Book = bookService.createBook(idempotencyKey, request)

}